[sBTC] Feat: Peg-in operation wire format

netrome opened the pull request:
This PR introduces an alpha version of the wire format for the Peg-in operation, intended for early end-to-end testing of sBTC. This version do not validate the peg wallet address, which is left to an external stacker application for the alpha test phase.

I've created #3494 as a follow-up task to add the validation of the peg wallet address once Epoch 3.0 and the updated `.pox` contract is ready.

igorsyl commented:
Adding @donpdonp, @sergey-shandar and @xoloki as reviewers to become familiar with this code.

jcnelson commented on src/chainstate/burn/db/sortdb.rs:
```diff
@@ -699,6 +725,20 @@ const SORTITION_DB_SCHEMA_4: &'static [&'static str] = &[
+        block_height INTEGER NOT NULL,
+        burn_header_hash TEXT NOT NULL,
+
+        address TEXT NOT NULL,
```
What's `address` here again?  Can you add a (sql) comment?  Thanks!

netrome replied:
Sure, it's the stacks address to receive the `sBTC`. I can add a comment.

Hmm, I think `recipient` would be a better field name throughout the code. I'll rename the field and add a SQL comment for it. Let me know if you have an even better name in mind.

netrome replied:
✅ Done.

jcnelson commented on src/chainstate/burn/operations/mod.rs:
```diff
@@ -308,6 +313,19 @@ pub struct DelegateStxOp {
 
+#[derive(Debug, PartialEq, Clone, Eq, Serialize, Deserialize)]
+pub struct PegInOp {
+    pub address: StacksAddress,
```
I know SIP-021's wire format appendix says that this is a Stacks address, but the text itself says that this can also be a smart contract address.  Can you make this the case?

Bytes 3-24 would be the address portion (1 byte version, 20 byte hash), and bytes 24-64 would be the name portion.

The name would be a smart contract name, so it would be between 5 and 40 bytes long.

netrome replied:
Sure thing, I'll update.

netrome replied:
I'll represent the smart contract name as a String for the alpha version, leaving room for a future discussion around improved type safety for contract addresses.

The reason is that I have not found any good existing types to represent smart contract addresses. The best type I know of is `clarity::vm::types::PrincipalData`, which uses `ContractName` to validate contract names. However, this would make `chainstate::burn::operations` depend on the `clarity` crate, which I would not like to do on a whim. Also, `ContractName` can introduce unintentional panics as observed in #3431, which would risk bringing down a running node on a malformed transaction.

I guess ideally it would be best to have type alongside `StacksAddress` in `stacks_common::types::chainstate` to represent contract addresses.

Let me know any thoughts on this.

netrome replied:
We discussed this on an sBTC working group meeting today. We're okay with depending on clarity types in `stacks_common::types::chainstate`, so I will change this ti use the `PrincipalData` type.

netrome replied:
Done :heavy_check_mark:

jcnelson commented on testnet/stacks-node/src/tests/neon_integrations.rs:
```diff
@@ -10417,3 +10420,93 @@ fn microblock_miner_multiple_attempts() {
+
+#[test]
+#[ignore]
+fn test_submit_and_observe_peg_in_request() {
```
While what you have here is great, I just wanted to let you know that there are lighter-weight operations available for testing burnchain operations.  For example, see `test_pox_btc_ops()` and `test_stx_transfer_btc_ops()` in `src/chainstate/coordinator/tests.rs`.

netrome replied:
Thanks! I'll study those test cases and see if it's worth complementing the existing tests with something more lightweight to further future-proof the code. Since the neon integration tests are slow, the more code paths we can cover in lighter tests the better.

netrome replied:
I added a test case for the peg-in operation in the `chainstate::coordinator::tests` now to complement the existing test cases.

netrome replied:
✅ Done

netrome commented:
@jcnelson FYI I'm not authorized to merge this PR. Can you add me as maintainer so I can merge once it's approved and CI is green?

jcnelson commented on clarity/src/vm/types/mod.rs:
```diff
@@ -1397,6 +1397,15 @@ impl From<StandardPrincipalData> for StacksAddress {
     }
 }
 
+impl From<PrincipalData> for StacksAddress {
```
Please remove this conversion. It should never be possible to convert a smart contract address into a standard address.

netrome replied:
Done.

jcnelson commented on src/chainstate/burn/db/sortdb.rs:
```diff
@@ -6127,6 +6224,50 @@ pub mod tests {
+            .expect("Failed to get peg-in ops from sortition DB");
+
+        assert_eq!(res_peg_ins.len(), 1);
+        assert_eq!(res_peg_ins[0].amount, 1337);
```
Can you add test coverage to verify that two different burnchain blocks can contain the same transaction (i.e. with the same txid), but that `SortitionDB::get_peg_in_ops()` will return the burnchain block-specific operation metadata for a given burnchain block hash?

netrome replied:
Done. I extended the test case to submit two peg in operations with different amounts and burn header hashes.

jcnelson commented on src/chainstate/burn/operations/peg_in.rs:
```diff
@@ -0,0 +1,451 @@
+            .map(TryInto::try_into)
+            .transpose()?;
+
+        println!("Contract name: {:?}", maybe_contract_name);
```
Please remove or use the logging facilities.

netrome replied:
Ouch, this was unintentional. Thanks for spotting.

jcnelson commented on src/chainstate/burn/operations/peg_in.rs:
```diff
@@ -0,0 +1,451 @@
+
+        let standard_principal_data = StandardPrincipalData(version, address_data);
+
+        let maybe_contract_name: Option<ContractName> = data
```
This code does not handle the case if there are trailing 0-bytes, or even the case where the `name` field is all 0's.  I think it should, because the space beyond the 64th byte can be used as a memo field for future extensions.  Instead, this code should just select out the leading non-zero bytes from the range 21 to 61, and try to decode _that_ into a ContractName.  If there are trailing 0's, it's fine -- they'll be ignored.  If the whole name field is 0, then there's no name.

Also, please don't use such long operator chains!  It's very hard to follow what this code is supposed to be doing.  Just do something like this:

```rust
let maybe_contract_name: Option<ContractName> = {
   // read to the first 0 byte
   let strlen = {
      let mut len = cmp::min(61, data.len());
      for i in 21..len {
         if data[i] == 0 {
            len = i;
            break;
          }
       }
       len
   };
      
   let name_bytes = data.get(21..strlen).unwrap_or(&[]);
   let name_string = std::str::from_utf8(&name_bytes).map_err(|_| ParseError::MalformedData)?.to_string();
   let name = ContractName::try_from(name_str).map_err(|_| ParseError::MalformedData)?;
   name
}
```

netrome replied:
Alright. I factored out some pieces of the suggestion to a `leading_non_zero_bytes` utility function and updated the functionality according to your description. It should be more readable now without the long operator chain. Let me know if you have any further opinions.

jcnelson commented on src/chainstate/burn/operations/peg_in.rs:
```diff
@@ -0,0 +1,451 @@
+            .expect("Any strictly positive amounts should be ok");
+    }
+
+    // TODO: Invalid contract name test
```
Please complete this task before merging.

netrome replied:
Done. I had already added the test case, but forgot to clean up the stale todo.

jcnelson commented on src/chainstate/coordinator/tests.rs:
```diff
@@ -3236,6 +3236,181 @@ fn test_stx_transfer_btc_ops() {
+        })
+        .count();
+
+    assert_eq!(total_number_of_peg_in_ops, 1);
```
Please also check that it's the one you expect.

netrome replied:
Done :heavy_check_mark:

jcnelson commented on testnet/stacks-node/src/tests/neon_integrations.rs:
```diff
@@ -10417,3 +10420,93 @@ fn microblock_miner_multiple_attempts() {
+#[test]
+#[ignore]
+fn test_submit_and_observe_peg_in_request() {
+    let receiver_stx_addr = StacksAddress::new(0, Hash160([0; 20]));
```
Please also add test coverage for receiving to a contract principal.

jcnelson commented:
A meta-comment.  When decoding a burnchain operation that is represented in fewer than 80 OP_RETURN bytes, you should save the remaining bytes into a `memo: Vec<u8>` and store them.  We use these "memo" bytes for in-band signaling between miners.

netrome commented:
> A meta-comment. When decoding a burnchain operation that is represented in fewer than 80 OP_RETURN bytes, you should save the remaining bytes into a `memo: Vec<u8>` and store them. We use these "memo" bytes for in-band signaling between miners.

Makes sense. I'll add it here. This memo field also seems to be missing in the `PreStxOp`, `StackStxOp` and `DelegateStxOp` operations. I created #3530 as a follow-up to add it for those operations too. @jcnelson feel free to close the issue if there's any reason these ops should not have the memo field.

jcnelson commented on src/chainstate/burn/db/sortdb.rs:
```diff
@@ -699,6 +731,23 @@ const SORTITION_DB_SCHEMA_4: &'static [&'static str] = &[
+        amount TEXT NOT NULL,
+        memo TEXT,
+
+        PRIMARY KEY(txid, sortition_id)
```
Hey, I know I said that this needs to be `(txid, sortition_id)` earlier, but it turns out I misspoke.  It should be `(txid, burn_header_hash)`.  It turns out that my PR #3528 is wrong as well.

jcnelson replied:
I left an explanation as to why #3528 is currently mis-designed here: https://github.com/stacks-network/stacks-blockchain/pull/3528#issuecomment-1408859705

netrome replied:
Alright, thanks for letting me know. I've updated the PR now using `(txid, burn_header_hash)` as the new primary key.

jcnelson commented on src/chainstate/burn/db/sortdb.rs:
```diff
@@ -699,6 +731,21 @@ const SORTITION_DB_SCHEMA_4: &'static [&'static str] = &[
 // update this to add new indexes
 const LAST_SORTITION_DB_INDEX: &'static str = "index_burn_header_hash_pox_valid";
```
Please change this constant to `index_peg_in_burn_header_hash`.

netrome replied:
Done ✅

jcnelson commented on src/chainstate/burn/operations/mod.rs:
```diff
@@ -308,6 +316,20 @@ pub struct DelegateStxOp {
+    pub recipient: PrincipalData,
+    pub peg_wallet_address: PoxAddress,
+    pub amount: u64,   // BTC amount to peg in, in satoshis
+    pub memo: Vec<u8>, // extra unused byte
```
This field can be more than 1 byte long. Please update the comment.

netrome replied:
Oops, thanks for spotting. Fixed!

jcnelson commented on src/chainstate/burn/operations/peg_in.rs:
```diff
@@ -0,0 +1,470 @@
+use stacks_common::codec::StacksMessageCodec;
```
Please add the copyright statement that we use in all the other source files to this one.

netrome replied:
Done.

jcnelson commented on src/chainstate/burn/operations/peg_in.rs:
```diff
@@ -0,0 +1,470 @@
+            if let Some(Some(recipient)) = tx.get_recipients().first() {
+                (recipient.amount, recipient.address.clone())
+            } else {
+                warn!("Invalid tx: Output 2 not provided");
```
N.B. This can also fail if the output is not recognized (i.e. is non-standard).

netrome replied:
Thanks for clarifying. I've updated the error message.

jcnelson commented on src/chainstate/burn/operations/peg_in.rs:
```diff
@@ -0,0 +1,470 @@
+            return Err(ParseError::MalformedData);
+        }
+
+        let version = *data.get(0).ok_or(ParseError::MalformedData)?;
```
Because we know that `data.len() >= 21`, you can instead use `.expect()` here instead of `.ok_or()`.

netrome replied:
Done.

jcnelson commented on src/chainstate/burn/operations/peg_in.rs:
```diff
@@ -0,0 +1,470 @@
+
+        Ok(())
+    }
+
```
Please add a docstring.

netrome replied:
Done.

jcnelson commented on src/chainstate/burn/operations/peg_in.rs:
```diff
@@ -0,0 +1,470 @@
+        Ok(())
+    }
+
+    fn leading_non_zero_bytes(data: &[u8], from: usize, to: usize) -> Option<&[u8]> {
```
A defensive check that `from < to` might be appropriate here.

netrome replied:
Makes sense. I added an assert since an error would indicate a severe bug in the code at which point a panic seems warranted. Let me know if you prefer a normal error here instead.

jcnelson commented on src/chainstate/burn/operations/peg_in.rs:
```diff
@@ -0,0 +1,470 @@
+    fn leading_non_zero_bytes(data: &[u8], from: usize, to: usize) -> Option<&[u8]> {
+        let end_of_non_zero_slice = {
+            let mut end = to.min(data.len());
+            for i in 21..end {
```
Bug -- should be `for i in from..end`, not 21.

netrome replied:
Thanks for spotting. Fixed.

jcnelson commented on src/chainstate/coordinator/tests.rs:
```diff
@@ -3236,6 +3236,187 @@ fn test_stx_transfer_btc_ops() {
+#[test]
+fn test_sbtc_peg_in_btc_op() {
+    let path = "/tmp/stacks-blockchain-sbtc-peg-in-btc-ops";
+    let _r = std::fs::remove_dir_all(path);
```
Can you just `.unwrap()` this so the test fails if the filesystem is inaccessible for this test?

netrome replied:
I can't simply unwrap, since that would cause the test to panic if the directory do not exist. I'm not a super fan of this pattern of ignoring results either but it's repeated in most test cases in this module so I'd recommend any changes to this pattern to be solved as a separate PR.

jcnelson commented on testnet/stacks-node/src/tests/neon_integrations.rs:
```diff
@@ -10417,3 +10422,151 @@ fn microblock_miner_multiple_attempts() {
+
+    next_block_and_wait(&mut btc_regtest_controller, &blocks_processed);
+
+    let parsed_peg_in_op_standard = {
```
This is fine for now, but this information should also be obtainable from the event observer.  Please open an issue so we can remember to go back and add it.  Thanks!

netrome replied:
Sure, I created #3554

jcnelson reviewed (commented):
This is mostly in good shape.  Just a few minor things to do and I'll approve.  Thanks @netrome!

netrome commented on src/chainstate/burn/db/sortdb.rs:
```diff
@@ -2869,6 +2918,18 @@ impl SortitionDB {
         Ok(())
     }
 
+    fn apply_schema_5(tx: &DBTx) -> Result<(), db_error> {
```
Alright, I've updated `is_db_version_supported_in_epoch` now to only allow schema 5 in Epoch 2.1

netrome commented on src/chainstate/burn/operations/peg_in.rs:
```diff
@@ -0,0 +1,470 @@
+    }
+
+    #[test]
+    fn test_parse_peg_in_should_succeed_given_a_contract_recipient() {
```
Done. I've added the memo to most of the other tests as well.

netrome commented:
> This is mostly in good shape. Just a few minor things to do and I'll approve. Thanks @netrome!

Thank you @jcnelson , I have addressed all your comments now. Let me know if everything looks good now so we can move this forward :rocket:

igorsyl commented:
In `stx_transfer_btc_integration_test`:
```
 #5 159.5 thread 'tests::neon_integrations::stx_transfer_btc_integration_test' panicked at 'unexpected btc transaction type', testnet/stacks-node/src/tests/neon_integrations.rs:1709:21
```

jcnelson commented:
cc @pavitthrap who touched this code path last

netrome commented:
@jcnelson @igorsyl why was `develop` merged into this PR? This PR is based on `next`, and merging in unrelated changes makes the version history unnecessarily complex and hard to follow. I'm happy to rebase if there's any new changes in `next` that we need to get this through. Otherwise, let's try to separate refactor noise from feature additions as already agreed.

netrome commented:
> In `stx_transfer_btc_integration_test`:
> 
> ```
>  #5 159.5 thread 'tests::neon_integrations::stx_transfer_btc_integration_test' panicked at 'unexpected btc transaction type', testnet/stacks-node/src/tests/neon_integrations.rs:1709:21
> ```

This is unrelated to the peg-in changes, and was introduced when `develop` was merged (which I've reverted now). This test failure needs to be addressed to unblock https://github.com/stacks-network/stacks-blockchain/pull/3564. FYI @pavitthrap

netrome commented:
Rebased on latest `next` and CI is green :heavy_check_mark:. @jcnelson I'd love it if you could have a look if this is okay to merge.

jcnelson reviewed (approved):
LGTM!  Great job on this @netrome!

netrome commented:
> LGTM! Great job on this @netrome!

Thank you! I'm not currently authorized to merge, can you press the magic button for me @jcnelson?

blockstack-devops commented:
This pull request has been automatically locked since there has not been any recent activity after it was closed. Please open a new issue for related bugs.
